### PR TITLE
Credit-infos

### PR DESCRIPTION
**Pull request by jdroenner**

### CODE REVIEW

**Comment by jdroenner** [comment]
[Quota im UI anzeigen](https://trello.com/c/aCVuoynv/1222-quota-im-ui-anzeigen)

File: projects/core/src/lib/users/user.service.ts [hunk 1]
```diff
@@ -120,6 +130,28 @@ export class UserService {
...
+                if (!backendInfo) {
+                    return false;
+                }
+                if (!backendInfo.features) {
+                    return false;
+                }
+                return backendInfo.features.includes(feature);
+            }),
+        );
+    }
```
Comment (ChristianBeilschmidt) [inline projects/core/src/lib/users/user.service.ts:153]: I'm not sure if we should rely on that.
If we do, we should not make it optional to output this.
But it was intended just for debugging purposes.

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        return 'primary';
+    }
+
+    protected hasQuota(): boolean {
+        return !!this.sessionQuota;
+    }
+
+    protected refreshQuota(): void {
+        this.updateQuota$.next(undefined);
```
Comment (ChristianBeilschmidt) [inline projects/core/src/lib/users/quota/quota-info/quota-info.component.ts:77]: For curiosity, can't you use send nothing if the type is void?

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.html [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        <span class="value">{{ sessionQuota?.available }} of {{ sessionQuota?.total }}</span>
+    </span>
+    <span class="pair">
+        <span class="label">Used credits (total)</span>
+        <span class="value">{{ sessionQuota?.used }} of {{ sessionQuota?.total }} </span>
+    </span>
+    <span class="pair">
+        <span class="label">Used credits (percent)</span>
+        <span class="value">{{ sessionQuota?.percentUsed }}%</span>
+        <mat-progress-bar mode="determinate" [value]="sessionQuota?.percentUsed" [color]="progressColor"></mat-progress-bar>
```
Comment (ChristianBeilschmidt) [inline projects/core/src/lib/users/quota/quota-info/quota-info.component.html:13]: grafik

This should be 8%.

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.html [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        <span class="label">Available credits</span>
+        <span class="value">{{ sessionQuota?.available }} of {{ sessionQuota?.total }}</span>
+    </span>
+    <span class="pair">
+        <span class="label">Used credits (total)</span>
+        <span class="value">{{ sessionQuota?.used }} of {{ sessionQuota?.total }} </span>
+    </span>
+    <span class="pair">
+        <span class="label">Used credits (percent)</span>
+        <span class="value">{{ sessionQuota?.percentUsed }}%</span>
```
Comment (ChristianBeilschmidt) [inline projects/core/src/lib/users/quota/quota-info/quota-info.component.html:12]: Maybe you can use the percentage pipe from Angular?

https://angular.io/api/common/PercentPipe

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.html [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<ng-container *ngIf="hasQuota()">
+    <span class="pair">
+        <span class="label">Available credits</span>
+        <span class="value">{{ sessionQuota?.available }} of {{ sessionQuota?.total }}</span>
```
Comment (ChristianBeilschmidt) [inline projects/core/src/lib/users/quota/quota-info/quota-info.component.html:4]: Can there be something available of …?

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                }
+                this.changeDetectorRef.markForCheck();
+            });
+
+        this.layerSubscription = this.projectService.getLayerStream().subscribe((_layer) => {
+            this.refreshQuota();
+        });
+        this.plotSubscription = this.projectService.getPlotStream().subscribe((_plot) => {
+            this.refreshQuota();
+        });
```
Comment (ChristianBeilschmidt) [inline projects/core/src/lib/users/quota/quota-info/quota-info.component.ts:53]: Why is this a good time to automatically update the quota?

If you want auto updates, you should think about attaching it to the same mechanism that updates the map?

Or… you could do it every 30 seconds or so.

What do you think?

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+    ngOnDestroy(): void {
+        this.sessionSubscription?.unsubscribe();
+        this.layerSubscription?.unsubscribe();
+        this.plotSubscription?.unsubscribe();
+    }
+
+    get progressColor(): string {
+        if (!this.sessionQuota) {
+            return 'gray';
```
Comment (ChristianBeilschmidt) [inline projects/core/src/lib/users/quota/quota-info/quota-info.component.ts:64]: I don't think this is valid.

I guess progressColor() should return `ThemePalette` instead.

`import {ThemePalette} from '@angular/material/core';`

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.html [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    </span>
+
+    <span class="sync">
+        <button mat-icon-button (click)="refreshQuota()"><mat-icon>sync</mat-icon></button>
+    </span>
+</ng-container>
+
+<ng-container *ngIf="!hasQuota()">
+    <p>No quota information available.</p>
+</ng-container>
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.html:23]: well if you use the component it should display something

File: projects/core/src/lib/users/user.service.ts (hunk 1 above)
Comment (jdroenner) [reply projects/core/src/lib/users/user.service.ts:153]: we should somehow expose backend capabilities

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.ts (hunk 6 above)
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.ts:53]: all of them are not a good trigger.

Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.ts:53]: what we would need is something that reloads if any layer or plot is loading (and when a layer is finished).

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.scss [hunk 9]
```diff
@@ -0,0 +1,37 @@
+.pair {
+    display: flex;
+    flex-direction: column;
+    justify-content: center;
+    align-items: start;
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.scss]: https://github.com/geo-engine/geoengine-ui/pull/586/commits/1171b374b96ee0e0af9906e6efba909a61d00362

File: projects/core/src/lib/users/user-session/user-session.component.scss [hunk 10]
```diff
@@ -0,0 +1,32 @@
+.pair {
+    display: flex;
+    flex-direction: column;
+    justify-content: center;
+    align-items: start;
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply projects/core/src/lib/users/user-session/user-session.component.scss]: [caf80ff](https://github.com/geo-engine/geoengine-ui/pull/586/commits/caf80ff70acc4e70cccb1ceffa616ed062bfe61e)

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.ts (hunk 2 above)
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.ts]: [959f18a](https://github.com/geo-engine/geoengine-ui/pull/586/commits/959f18a6b945739a71e8136c46b4921caa6c5bd2)

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.html (hunk 3 above)
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.html]: [9cbccf5](https://github.com/geo-engine/geoengine-ui/pull/586/commits/9cbccf56c567b103e556ba219e6260ed50dd9b63)

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.html (hunk 5 above)
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.html]: [825d5aa](https://github.com/geo-engine/geoengine-ui/pull/586/commits/825d5aa91b42a58563a534bb57f528a8d07cf0e6)

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.ts (hunk 7 above)
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.ts]: [58b8758](https://github.com/geo-engine/geoengine-ui/pull/586/commits/58b875872fb50dde85173efc2b933ff9ade19431)

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.html (hunk 4 above)
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.html]: [9cbccf5](https://github.com/geo-engine/geoengine-ui/pull/586/commits/9cbccf56c567b103e556ba219e6260ed50dd9b63)

File: projects/core/src/lib/users/user.service.ts (hunk 1 above)
Comment (jdroenner) [reply projects/core/src/lib/users/user.service.ts]: i removed it [e46564b](https://github.com/geo-engine/geoengine-ui/pull/586/commits/e46564b203836f6ce2baf1a8d4eec605fc520333)

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.ts (hunk 2 above)
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.ts]: i moved that to the user service:
[e46564b](https://github.com/geo-engine/geoengine-ui/pull/586/commits/e46564b203836f6ce2baf1a8d4eec605fc520333)

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.html (hunk 3 above)
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.html]: i removed all the percents
[e46564b](https://github.com/geo-engine/geoengine-ui/pull/586/commits/e46564b203836f6ce2baf1a8d4eec605fc520333)

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.html (hunk 5 above)
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.html]: there is no longer a total
[e46564b](https://github.com/geo-engine/geoengine-ui/pull/586/commits/e46564b203836f6ce2baf1a8d4eec605fc520333)

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.ts (hunk 6 above)
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.ts]: it uses a timer now

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.ts (hunk 7 above)
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.ts]: the progessbar was removed [e46564b](https://github.com/geo-engine/geoengine-ui/pull/586/commits/e46564b203836f6ce2baf1a8d4eec605fc520333)

**Comment by jdroenner** [comment]
we had a discussion if a total makes sense and then removed the total and percent stuff.
[e46564b](https://github.com/geo-engine/geoengine-ui/pull/586/commits/e46564b203836f6ce2baf1a8d4eec605fc520333)

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.html:18]: This PR only adds the information to the  OIDC-Login-Component. I guess this kind of login also only makes sense with the pro version?

File: projects/core/src/lib/users/quota/quota-info/quota-info.component.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    ngOnInit(): void {
+        this.sessionQuotaSubscription = this.userService.getSessionQuotaStream().subscribe((quota) => {
+            this.sessionQuota = quota;
+            this.changeDetectorRef.detectChanges();
+        });
+
+        setInterval(() => {
+            this.refreshQuota();
+        }, 3000);
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.ts]: [9ef8fba](https://github.com/geo-engine/geoengine-ui/pull/586/commits/9ef8fbaa9bb8dc92f1600e9b3aa79a74a58b80d5)

[the remark this answers is not in the document]
Comment (jdroenner) [reply projects/core/src/lib/users/quota/quota-info/quota-info.component.ts]: [clear inteval (30sec)](https://github.com/geo-engine/geoengine-ui/pull/586/commits/9ef8fbaa9bb8dc92f1600e9b3aa79a74a58b80d5)
[9ef8fba](https://github.com/geo-engine/geoengine-ui/pull/586/commits/9ef8fbaa9bb8dc92f1600e9b3aa79a74a58b80d5)

[use rxjs timer](https://github.com/geo-engine/geoengine-ui/pull/586/commits/33aa5d7747c680fd4bb60d7bcaaebedb1382ef4b)
[33aa5d7](https://github.com/geo-engine/geoengine-ui/pull/586/commits/33aa5d7747c680fd4bb60d7bcaaebedb1382ef4b)

**Review by ChristianBeilschmidt — APPROVED** [review]

**Comment by ChristianBeilschmidt** [comment]
bors merge

**Pull request merged**